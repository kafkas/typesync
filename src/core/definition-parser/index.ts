export * from './definition-parser.js';
