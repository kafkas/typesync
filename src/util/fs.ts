import { existsSync, mkdirSync, readdirSync, statSync } from 'node:fs';
import { writeFile as _writeFile } from 'node:fs/promises';
import 'node:os';
import { dirname } from 'node:path';

import { extractErrorMessage } from './extract-error-message';

/**
 * Checks if the specified directory is empty. Creates the directory if it can be created.
 * Otherwise throws.
 */
export function validateEmptyDir(pathToDir: string) {
  try {
    const stats = statSync(pathToDir);

    if (!stats.isDirectory()) {
      throw new Error(`The path ${pathToDir} is non-empty.`);
    }

    const isEmpty = readdirSync(pathToDir).length === 0;

    if (!isEmpty) {
      throw new Error(`The directory ${pathToDir} is non-empty.`);
    }
  } catch {
    try {
      mkdirSync(pathToDir, { recursive: true });

      const isEmpty = readdirSync(pathToDir).length === 0;

      if (!isEmpty) {
        throw new Error(`The directory ${pathToDir} is non-empty.`);
      }
    } catch (e) {
      throw new Error(`Unexpected error in mkdirSync(): ${extractErrorMessage(e)}`);
    }
  }
}

export async function writeFile(
  path: string,
  data: Parameters<typeof _writeFile>[1],
  opts?: Parameters<typeof _writeFile>[2]
) {
  ensureFileDirectoryExists(path);
  await _writeFile(path, data, opts);
}

function ensureFileDirectoryExists(pathToFile: string) {
  const dirName = dirname(pathToFile);
  if (existsSync(dirName)) {
    return;
  }
  ensureFileDirectoryExists(dirName);
  mkdirSync(dirName);
}
