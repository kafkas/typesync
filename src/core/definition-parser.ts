import { readFileSync } from 'node:fs';
import { parse as parseYaml } from 'yaml';
import { z } from 'zod';

import { definition } from '../definition';
import { DefinitionNotValidError, DefinitionNotValidYamlError } from '../errors';
import type { DefinitionParser, Logger } from '../interfaces';
import { schema } from '../schema';
import { extractErrorMessage } from '../util/extract-error-message';

class DefinitionParserImpl implements DefinitionParser {
  public constructor(private readonly logger: Logger) {}

  public parseDefinition(pathToDefinition: string): schema.Schema {
    const definitionJson = this.parseYamlFileAsJson(pathToDefinition);
    const aliasNames = this.extractAliasModelNames(definitionJson);
    const definitionSchema = definition.schemas.definition(aliasNames);
    const parsedDefinition = this.parseDefinitionWithSchema(definitionJson, definitionSchema);
    return schema.createFromDefinition(parsedDefinition);
  }

  private parseYamlFileAsJson(pathToFile: string): unknown {
    try {
      const yamlContent = readFileSync(pathToFile).toString();
      return parseYaml(yamlContent, { strict: true });
    } catch (e) {
      this.logger.error(extractErrorMessage(e));
      throw new DefinitionNotValidYamlError(pathToFile);
    }
  }

  private extractAliasModelNames(definitionContentJson: unknown) {
    const looseDefinitionSchema = z.record(
      z.object({
        type: z.enum(['document', 'alias']),
      })
    );
    const parsedDefinition = this.parseDefinitionWithSchema(definitionContentJson, looseDefinitionSchema);
    return Object.entries(parsedDefinition)
      .filter(([, model]) => model.type === 'alias')
      .map(([name]) => name);
  }

  private parseDefinitionWithSchema<T>(content: unknown, zodSchema: z.Schema<T>) {
    const parseRes = zodSchema.safeParse(content);

    if (!parseRes.success) {
      const { error } = parseRes;
      const issue = error.issues[0];
      if (issue) {
        const { message } = issue;
        const path = issue.path.join('.');
        throw new DefinitionNotValidError(path, message);
      } else {
        throw new DefinitionNotValidError('unknown', 'An unexpected invalid value.');
      }
    }

    return parseRes.data;
  }
}

export function createDefinitionParser(logger: Logger): DefinitionParser {
  return new DefinitionParserImpl(logger);
}
