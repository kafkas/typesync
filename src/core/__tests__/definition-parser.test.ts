import { globSync } from 'glob';
import { resolve } from 'path';

import { getDirName } from '../../util/fs.js';
import { createDefinitionParser } from '../definition-parser.js';

describe('definition-parser', () => {
  it('correctly parses definition from YAML files', () => {
    const parser = createDefinitionParser();
    const definitionGlobPattern = resolve(getDirName(import.meta.url), `./definitions/yaml/*.yml`);
    const filePaths = globSync(definitionGlobPattern);
    const definition = parser.parseDefinition(filePaths);
    expect(definition).toMatchSnapshot();
  });

  it('correctly parses definition from JSON files', () => {
    const parser = createDefinitionParser();
    const definitionGlobPattern = resolve(getDirName(import.meta.url), `./definitions/json/*.json`);
    const filePaths = globSync(definitionGlobPattern);
    const definition = parser.parseDefinition(filePaths);
    expect(definition).toMatchSnapshot();
  });

  it(`ignores the '$schema' field in definition files`, () => {
    const parser = createDefinitionParser();
    const definitionGlobPattern = resolve(getDirName(import.meta.url), `./definitions/with-$schema-key.json`);
    const filePaths = globSync(definitionGlobPattern);
    const definition = parser.parseDefinition(filePaths);
    expect(definition).toMatchSnapshot();
  });

  it(`throws if the file contains an incorrect field`, () => {
    const parser = createDefinitionParser();
    const definitionGlobPattern = resolve(getDirName(import.meta.url), `./definitions/bad-field.json`);
    const filePaths = globSync(definitionGlobPattern);
    expect(() => parser.parseDefinition(filePaths)).toThrow();
  });
});
