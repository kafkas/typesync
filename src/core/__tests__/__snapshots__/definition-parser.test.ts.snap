// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definition-parser correctly parses definition from JSON files 1`] = `
{
  "Address": {
    "model": "alias",
    "type": {
      "fields": {
        "city": {
          "type": "string",
        },
        "street": {
          "type": "string",
        },
        "zip_code": {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "Cat": {
    "model": "alias",
    "type": {
      "fields": {
        "lives_left": {
          "type": "int",
        },
        "name": {
          "type": "string",
        },
        "type": {
          "type": {
            "type": "literal",
            "value": "cat",
          },
        },
      },
      "type": "object",
    },
  },
  "Dog": {
    "model": "alias",
    "type": {
      "fields": {
        "breed": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "type": {
          "type": {
            "type": "literal",
            "value": "dog",
          },
        },
      },
      "type": "object",
    },
  },
  "Pet": {
    "model": "alias",
    "type": {
      "type": "union",
      "variants": [
        "Cat",
        "Dog",
      ],
    },
  },
  "UserProfile": {
    "docs": "User profile model",
    "model": "document",
    "type": {
      "fields": {
        "address": {
          "type": "Address",
        },
        "age": {
          "type": "int",
        },
        "area_code": {
          "type": {
            "type": "literal",
            "value": 34,
          },
        },
        "bio": {
          "type": {
            "type": "union",
            "variants": [
              "nil",
              "string",
            ],
          },
        },
        "created_at": {
          "optional": true,
          "type": "timestamp",
        },
        "favorite_numbers": {
          "type": {
            "elementType": "int",
            "type": "list",
          },
        },
        "is_active": {
          "type": "boolean",
        },
        "location": {
          "type": {
            "elements": [
              "int",
              "int",
            ],
            "type": "tuple",
          },
        },
        "metadata": {
          "type": {
            "type": "map",
            "valueType": "unknown",
          },
        },
        "pets": {
          "type": {
            "elementType": "Pet",
            "type": "list",
          },
        },
        "petsById": {
          "type": {
            "type": "map",
            "valueType": "Pet",
          },
        },
        "role": {
          "type": "UserRole",
        },
        "username": {
          "type": "Username",
        },
      },
      "type": "object",
    },
  },
  "UserRole": {
    "docs": "Represents a user role",
    "model": "alias",
    "type": {
      "members": [
        {
          "label": "ADMIN",
          "value": "admin",
        },
        {
          "label": "USER",
          "value": "user",
        },
        {
          "label": "GUEST",
          "value": "guest",
        },
      ],
      "type": "enum",
    },
  },
  "Username": {
    "model": "alias",
    "type": "string",
  },
}
`;

exports[`definition-parser correctly parses definition from YAML files 1`] = `
{
  "Address": {
    "model": "alias",
    "type": {
      "fields": {
        "city": {
          "type": "string",
        },
        "street": {
          "type": "string",
        },
        "zip_code": {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "Cat": {
    "model": "alias",
    "type": {
      "fields": {
        "lives_left": {
          "type": "int",
        },
        "name": {
          "type": "string",
        },
        "type": {
          "type": {
            "type": "literal",
            "value": "cat",
          },
        },
      },
      "type": "object",
    },
  },
  "Dog": {
    "model": "alias",
    "type": {
      "fields": {
        "breed": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "type": {
          "type": {
            "type": "literal",
            "value": "dog",
          },
        },
      },
      "type": "object",
    },
  },
  "Pet": {
    "model": "alias",
    "type": {
      "type": "union",
      "variants": [
        "Cat",
        "Dog",
      ],
    },
  },
  "UserProfile": {
    "docs": "User profile model",
    "model": "document",
    "type": {
      "fields": {
        "address": {
          "type": "Address",
        },
        "age": {
          "type": "int",
        },
        "area_code": {
          "type": {
            "type": "literal",
            "value": 34,
          },
        },
        "bio": {
          "type": {
            "type": "union",
            "variants": [
              "nil",
              "string",
            ],
          },
        },
        "created_at": {
          "optional": true,
          "type": "timestamp",
        },
        "favorite_numbers": {
          "type": {
            "elementType": "int",
            "type": "list",
          },
        },
        "is_active": {
          "type": "boolean",
        },
        "location": {
          "type": {
            "elements": [
              "int",
              "int",
            ],
            "type": "tuple",
          },
        },
        "metadata": {
          "type": {
            "type": "map",
            "valueType": "unknown",
          },
        },
        "pets": {
          "type": {
            "elementType": "Pet",
            "type": "list",
          },
        },
        "petsById": {
          "type": {
            "type": "map",
            "valueType": "Pet",
          },
        },
        "role": {
          "type": "UserRole",
        },
        "username": {
          "type": "Username",
        },
      },
      "type": "object",
    },
  },
  "UserRole": {
    "docs": "Represents a user role",
    "model": "alias",
    "type": {
      "members": [
        {
          "label": "ADMIN",
          "value": "admin",
        },
        {
          "label": "USER",
          "value": "user",
        },
        {
          "label": "GUEST",
          "value": "guest",
        },
      ],
      "type": "enum",
    },
  },
  "Username": {
    "model": "alias",
    "type": "string",
  },
}
`;

exports[`definition-parser ignores the '$schema' field in definition files 1`] = `
{
  "Username": {
    "model": "alias",
    "type": "string",
  },
}
`;
